agent/agentserver/server_test.go

```diff
@@ -276,13 +277,82 @@ func TestReadinessCheckHandler(t *testing.T) {
 	}
 }
 
+func TestReadinessCheckHandlerCache(t *testing.T) {
+	for _, tc := range []struct {
+		desc              string
+		firstCallErr      error
+		readinessCacheTTL time.Duration
+		waitInvalidation  bool
+	}{
+		{
+			desc:              "call 1 succeeds, so second call automatically succeeds",
+			firstCallErr:      nil,
+			readinessCacheTTL: 10 * time.Minute,
+			waitInvalidation:  false,
+		},
+		{
+			desc:              "call 1 fails, so second call performs checks",
+			firstCallErr:      errors.New("test-error"),
+			readinessCacheTTL: 10 * time.Minute,
+			waitInvalidation:  false,
+		},
+		{
+			desc:              "call 1 succeeds, but cache becomes invalid, so second call performs checks",
+			firstCallErr:      nil,
+			readinessCacheTTL: 10 * time.Minute,
+			waitInvalidation:  true,
+		},
+		{
+			desc:              "call 1 succeeds, but caching is disabled, so second call performs checks",
+			firstCallErr:      nil,
+			readinessCacheTTL: 0,
+			waitInvalidation:  false,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			require := require.New(t)
+			mocks, cleanup := newServerMocks(t)
+			defer cleanup()
+
+			mocks.sched.EXPECT().Probe().Return(tc.firstCallErr).Times(1)
```

Comment:
Instead of parameterizing the return values of the mock calls and other behavior you can have a `setupFn(*mocks)` parameter in the test struct. Then for each test the mocks are set up as you need them. The test body will not have any EXPECT() calls, all of them will just be in the setupFn.

This is nice because there's less indirection and the test cases are no coupled with each other - it's easier to set up different types of mock behaviors.

Reply:
Done.